src/native/minipal/memory.c

```diff
@@ -0,0 +1,235 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#include <assert.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <sys/mman.h>
+#include <unistd.h>
+#include <minipal/memory.h>
+#include <minipal/thread.h>
+
+#ifdef __APPLE__
+#include <mach/thread_state.h>
+
+#define CHECK_MACH(_msg, machret) do {                                      \
+        if (machret != KERN_SUCCESS)                                        \
+        {                                                                   \
+            char _szError[1024];                                            \
+            snprintf(_szError, ARRAY_SIZE(_szError), "%s: %u: %s", __FUNCTION__, __LINE__, _msg);  \
+            mach_error(_szError, machret);                                  \
+            abort();                                                        \
+        }                                                                   \
+    } while (false)
+
+#endif // __APPLE__
+
+#ifndef TARGET_WINDOWS
+#ifdef __linux__
+#include <linux/membarrier.h>
+#include <sys/syscall.h>
+#define membarrier(...) syscall(__NR_membarrier, __VA_ARGS__)
+#elif HAVE_SYS_MEMBARRIER_H
+#include <sys/membarrier.h>
+#ifdef TARGET_BROWSER
+#define membarrier(cmd, flags, cpu_id) 0 // browser/wasm is currently single threaded
+#endif
+#endif
+
+static bool CanFlushUsingMembarrier()
+{
+#if defined(__linux__) || HAVE_SYS_MEMBARRIER_H
+
+#ifdef TARGET_ANDROID
+    // Avoid calling membarrier on older Android versions where membarrier
+    // may be barred by seccomp causing the process to be killed.
+    int apiLevel = android_get_device_api_level();
+    if (apiLevel < __ANDROID_API_Q__)
+    {
+        return false;
+    }
+#endif
+
+    // Starting with Linux kernel 4.14, process memory barriers can be generated
+    // using MEMBARRIER_CMD_PRIVATE_EXPEDITED.
+
+    int mask = membarrier(MEMBARRIER_CMD_QUERY, 0, 0);
+
+    if (mask >= 0 &&
+        mask & MEMBARRIER_CMD_PRIVATE_EXPEDITED &&
+        // Register intent to use the private expedited command.
+        membarrier(MEMBARRIER_CMD_REGISTER_PRIVATE_EXPEDITED, 0, 0) == 0)
+    {
+        return true;
+    }
+#endif
+
+    return false;
+}
+
+//
+// Tracks if the OS supports FlushProcessWriteBuffers using membarrier
+//
+static int s_flushUsingMemBarrier = 0;
+
+// Helper memory page used by the FlushProcessWriteBuffers
+static uint8_t* g_helperPage = 0;
+
+// Mutex to make the FlushProcessWriteBuffersMutex thread safe
+static pthread_mutex_t g_flushProcessWriteBuffersMutex;
+
+size_t g_pageSizeUnixInl = 0;
+
+bool minipal_initialize_flush_process_write_buffers()
+{
+    int pageSize = sysconf( _SC_PAGE_SIZE );
+
+    g_pageSizeUnixInl = (size_t)((pageSize > 0) ? pageSize : 0x1000);
```

Comment:
`g_pageSizeUnixInl` has nothing to do flush process write buffers. It is odd to initialize it in `minipal_initialize_flush_process_write_buffers`.

We may want to avoid touching page size in this PR, and just

Reply:
this was just convenient place to use, which is called during initialization. we could introduce minipal initialize to have better place for it. the idea was to deduplicate page size as well. otoh it is small part so I will leave it out then.